Create initial readme for Rust course and outline lesson plan

brson opened the pull request:
This PR describes a Rust training course called "Practical Networked Applications in Rust", and lays out a draft lesson plan.

[Rendered README](https://github.com/brson/talent-training/tree/master/rust).

[Rendered lesson plan](https://github.com/brson/talent-training/blob/master/rust/plan.md).

The flow of this course is curious, with topics derived from the needs of the hands-on projects and the hands-on projects designed to teach useful and interesting topics, leaving some important topics not covered (at least yet). I believe the lesson plan proceeds quite differently from typical Rust intro/beginner material. I hope that the documentation conveys the rationale sufficiently.

It is not quite finished, but I think there's enough here to understand what it's about, enough to review and critique.

The structure of this course is inspired by the [MIT 5.824 distributed systems course](https://pdos.csail.mit.edu/6.824/schedule.html), which we are also adapting to Rust as a follow-on course to this one. I intend that the two courses ultimately share a common structure.

The lesson plan is intended to be an index, with links to the full lessons and projects. The descriptions within the lesson plan of the goals, topics etc are intended as a guide to writing the full lessons and project descriptions. Once those are written we might reduce the descriptive content in the lesson plan so that it's easier for the reader to take in the full lesson/project progression by skimming.

*I very much want feedback. Are the goals and non-goals reasonable? Is the project interesting? Does the lesson plan make sense? Are the topics interesting? Within the given lesson plan, and considering the projects being implemented, what other topics should be included? As a new Rust programmer, does the content here excite you to take the course? What changes would make the course more exciting?*

I'm also looking for people interested in helping build this course, since there's quite a lot of material to cover. If this looks like something you would like to participate in, please speak up.

Note that there is a file here called [notes.md](https://github.com/brson/talent-training/blob/master/rust/notes.md) that contains notes I've taken about this course. It includes a list of potential subjects and links to other courses and draft training material. I intend the material to take inspiration, if not content, from https://github.com/ferrous-systems/rust-three-days-course, https://github.com/nrc/talks, and other existing training material. We'll do a survey of existing training material soon, with goals of understanding how this course can provide unique value, and how other courses progress from topic to topic.

I was hoping that this PR would fill out some of the material's structure, like how the slides and text are organized, an example full project description, how the projects' tests work; but I haven't made it that far yet.

cc @siddontang PTAL. I hope that the projects are aligned with your expectations. I suspect the volume of lesson material here is more ambitious than you expected, and is very probably more than can be completed in Q1. We can deprioritize certain lessons for an MVP. Give me all the feedback you've got. We haven't discussed how this course is licensed or who owns it, but I've preemptively put a CC-BY notice at the bottom of the readme so that there's some intent conveyed as to the license.

cc @Hoverbear @breeswish @overvenus @rleungx @Hijiao @huachaohuang @nrc if you have the time I'd be grateful for your feedback.

cc @skade I'd be grateful for any review you can offer. Hoverbear and I have been inspired by your rust-three-days-course and you've got lots of experience in Rust training.

nrc commented:
This looks like a really interesting course and a great way to learn Rust, I'm probably a bit biased because it fits in very closely with my current interests :-) I think the selection of topics is ambitious, but has great coverage.

I also have questions!

What is the value that we want to bring to students by taking this course, vs just reading the book and similar resources? (IOW, why make this whole thing vs just a reading list?). I think there are several possible benefits, but they should be listed and prioritised to help provide focus (there is a risk with this project, I think, that we do a lot of work but only provide a little value over existing resources, we need to be very vigilant to avoid that).

What kind of internals stuff will you cover and what is the goal of teaching it? Although I'm always a fan of such material, I have found that unless there is a very concrete goal it is not good value for most students.

I wonder if it is better to separate out the specific networking stuff from the general intermediate Rust stuff (e.g., build scripts, internals)? ISTM that the two audiences might be quite different.

Where do we expect this to live long term? Is this going to be primarily maintained by PingCAP or donated to the wider Rust community?

It sounds like the audience for this repo is instructors rather than students, do we also plan to make a complete student-targetted version (e.g., with videos of talks)? Do we plan to provide material/guidance for a competent Rust programmer to up-skill so that they can teach this course?

liufuyang commented:
`As a new Rust programmer`, the (future) content here excites me a lot to take the course. Months ago I took a look at 5.824 but I didn't spend too much time on it. Firstly because I could not find good docs/videos on the theory part (I like those courses format in platform such as coursera/edx). Secondly, as I was learning Rust at that moment, the lab of 5.824 is in Golang, so I didn't had enough energy/time to learn a new language nor finish some parts of the course.

I also feel, the content of 5.824 is a lot. For people like me (who might have a full time job or have a family) to learn something big during free time, it might be a good idea to split the big course into a few separate small parts. Like this one here: start from a memory kv store and the in the end make it a distributed system over network. I think the scope sounds very fun and challenging as well. It could be a very good start.

As nrc asked about the goal of the course, from my perspective, **the goal** of a student like me to learn this course could be: **to be able to apply a position at PingCap(or similar type of positions at other companies) afterwards, or to have a good feeling for better understanding distributed systems (kv stores) and feel much more experienced with Rust afterwards.**

About the content/format of the course, some people might like reading docs to learn, but I do like to learn via short videos. I don't know, but perhaps we can even later make this course on platforms such as coursera/edx/udemy? Making good teaching videos indeed takes lots of extra time and effort. But the good part of publishing those courses on those platforms might also make the content creators getting economic benefit. And the course can easily reach out to a border group of audiences via a course platform. But anyway, it is probably too early to start thinking in this direction. It is just from my own feeling, as I have had quite a lot courses on those platforms, it would be super awesome to have a course there about distributed kv stores in Rust and presented in the format of online courses (with good videos lectures and good homeworks, for example like the course called CS50, or the deeplearning.ai courses, which are all truly amazing).

I also hope in the future I can know these Rust topics well and being able to teach some other new comers. Making teaching materials and giving lectures are very fun and fulfilling things to do. But it is also challenging and requiring big effort to make it good. I am far from having enough skills/experiences to be helpful but I am looking forward to it :) 

Anyway, I hope some of my thoughts above do make some sense to you. Thanks for starting making a course like this. I will keep an eye on how it goes and start learning it as long as it is doable for me :)

siddontang commented on rust/README.md:
```diff
@@ -0,0 +1,135 @@
+The final project in this course is a networked, persistent [key-value store]
+with multithreading, asynchronous I/O (via [tokio] and [tower]), a
+high-performance network protocol ([GRPC] via [prost] and [tower-grpc]),
+backed by a simple [LSM-tree] that the student will write themselves.
```
oh, building an LSM tree is interesting but hard work.
I think this is also a challenge for us to teach students something about how to use Rust to write this :-)

brson replied:
Agreed it's hard, and I'm on the fence about it. It would definitely be an "extra credit" type thing. And I'm not necessarily thinking a full LSM-tree. More like "when the WAL hits a threshold, dump it to a binary-searchable file; when too many files accumulate merge them", but even that is quite a lot of work. The thing I like about it is that it exercises file I/O apis more than simple serialization and that it exercises an algorithm using those APIs.

I'll just remove this wording, maybe leave hints of it in the lesson plan, and see as we go if it's just too hard to ask students to think about.

siddontang commented on rust/README.md:
```diff
@@ -0,0 +1,135 @@
+- Structuring and maintaining Rust programs
+- Fun and foolproof parallel programming
+- Asyncronous programming with [futures] and [tokio]
+- High-performance networking with [GRPC] and [prost]
```
do we need to use tower-grpc or grpc-rs?

brson replied:
Either is fine with me. I put tower-grpc there since that's what we think we're heading to. Since it sounds like you prefer grpc-rs I'll use that and take out mentions of tower.

siddontang replied:
no, I prefer pure Rust tower-grpc :-)

siddontang commented:
Hi @nrc 

Now, you can think the talent training is maintained by PingCAP, so here what we want most is to train students and let them contribute to PinCAP. That is why we have a concrete goal here, we don't want to train a student just know Rust completely(I think there are many materials do this), but to train a student who later can use Rust to build a high-performance service, know something deeply about file I/O, network, concurrency. 

Our talent plan contains multi parts, Rust is only one part, after Rust, we will provide a distributed system course which trains students how to write a fault-tolerant, horizontal-scalability key-value database (Aha, a simple TiKV). And we also contain other parts like TiDB, Kurbenetes, etc. 

Btw, I think Rust is fine for building service as Go does, but sadly, there is no proper course does this, and we want to fill this gap. So maybe this Rust course looks concrete, I still think many people can get benefit from it.

siddontang commented:
@brson 

I think now we have a big ambition now. :-) This course may be hard for students to finish it in a short time. Maybe we need to separate it into different levels like basic and advanced levels. And at first, I think we can only provide a basic level to let students finish all in less than one month. 

Another thing I worry about is that this course may be also difficult for us to make the materials, especially for giving the tests, I think we need the communities to help us :-)

You worried that the course can't be given in Q1, don't worry, I think we can make the materials at first.

skade commented:
Hi! Thanks for thinking of me. I like what I currently see. First of all, I have to mention that it is perfectly feasible to run a `async Rust` course in a day (I've done so) and I don't think this is overambitious :).

I have 2 big questions:
* What is the exact audience this course caters at? How much previous programming knowledge should people have?
* Do you think there's a chance to produce an example program that can be used for teaching concurrency styles that can be shared? I've had great success with sketch programs of certain patterns, but haven't quite found a handle of turning this into a shareable artefact.

I'd be very interested in being low-key involved.

brson commented:
Thanks for the great feedback all ❤️ .

> What is the value that we want to bring to students by taking this course, vs just reading the book and similar resources? (IOW, why make this whole thing vs just a reading list?)

@nrc This is one of the big, obvious questions, one that I struggle with, and one of the biggest risks of the project I think. @siddontang addressed this a bit. Some of my ideas: this beelines to writing a pretty serious application, using the tech stack that TiKV uses. The book _also_ has a networking project in it and it's well described; but it uses std APIs, where this course leads the user toward the crates they might use for writing a more serious project. So this course is _directed_ toward a purpose, cutting out subjects that don't serve that purpose, whereas The Book and other such material is pretty directionless. Hopefully that produces a more engaging learning experience.

There are definitely some insights we can provide that aren't in the primary sources - e.g. in the last talk I gave on error handling I went pretty deep into `fn main() -> Result`, why production software probably shouldn't use it, diving into its implementation to explain its limitations. In this course we would expect the student to do robust error reporting as though it were a real product.

I think there's definitely a risk that, for some of these lessons, we'll gather the readings, try to write the material, and not be able to come up with much to say in addition. I wouldn't be afraid to just cut out the extra material, and only list the readings, though I'm hopeful we can come up with _some_ additional insight that serves the student, ideally to help them complete the project. This is one of the reasons the readme promises deep-dives - on almost any subject there's something more in-depth, and hopefully interesting, to be said than in the primary sources, though I did note your reservations about their usefulness.

I will put some more effort into expressing this in the readme.

> What kind of internals stuff will you cover and what is the goal of teaching it? Although I'm always a fan of such material, I have found that unless there is a very concrete goal it is not good value for most students.

In general, my goals with deep dives are to: show that reading std (or other code) is approachable, and a great way to learn; give insight into how and why things work the way they do, which can lead to greater understanding of how to use and intuit about features; just provide interesting info that they wouldn't get elsewhere.

It's definitely one of my biases, but I really like taking the lid off of features and explaining them down to their guts. I _think_ it's a good way to provide exposure to many things at once.

Examples.

I mentioned the `fn main() -> Result`, which includes exploring the `Termination` trait, which I think provides insight into the operation of traits (which is not otherwise covered in this course), explains the limitations of Rust's exit reporting, and shows that reading the standard library is easy. It's also potentially an opportunity to demystify the runtime startup process, since it's easy to trace the entry point forward to the error reporting code.

I've also previously talked about, and probably would in this course, how `?` works. Again I think it helps to demystify a piece of the language that looks like magic. It's also an opportunity to dig into the compiler and discuss desugaring, show how so much of the language they interact with is just surface syntax, which I think would be interesting to most Rust programmers. It is definitely a bit of a tangent from the purpose of building an application. But that is the kind of extra value I can see us adding.

I also have listed a dive into the formatting mechanisms, which would provide some exposure to macro internals, which this course otherwise doesn't touch; also the opportunity to dig into the compiler, show where the lines blur between the libraries and the compiler; explain how it's possible that Rust statically checks its format strings. I haven't thought this through too hard, and if it doesn't pan out the whole lesson could be cut without any loss.

Those are the only examples I have off the top of my head.

> I wonder if it is better to separate out the specific networking stuff from the general intermediate Rust stuff (e.g., build scripts, internals)? ISTM that the two audiences might be quite different.

It might be. Doing so might also strip the lessons of anything interesting.

The build script subject at least I put in there because it feeds into the project - need to run build scripts to compile protocol buffers. It is true that to provide value, I would intend to go much deeper into build scripts than necessary for the project. I do see build scripts as a subject that is not that well covered elsewhere, and suspect we could come up with interesting practical advice and demonstrations on the subject.

> Where do we expect this to live long term? Is this going to be primarily maintained by PingCAP or donated to the wider Rust community?

IMO it would be great if it was bigger than just PingCAP, though I don't imagine that happening until the project has begun to prove its worth. If there became interest, donating it to the community in some way would certainly be a good gesture, and an opportunity to bring in contributors. I do know that there's some desire to use this to more strongly associate PingCAP with Rust, so there's a tension there - like, for a live presentation we might expect to use slide decorations with PingCAP branding. PingCAP might ultimately be unwilling to part with the project, but I can't speak to that.


> It sounds like the audience for this repo is instructors rather than students, do we also plan to make a complete student-targetted version (e.g., with videos of talks)?

I think it's the opposite. Like the online coursework for the MIT course is followed by students and non-MIT learners, I expect more people to work through this material online than live (I'm actually concerned that the structure is not conducive to the kind of live training that we typically think of, since it has per-requisite readings and difficult hands-on projects). I am ambitiously hoping to structure it so that the decks are accompanied by text (which also feed into the speakernotes) in a way that the lessons basically read like a book chapter.

> Do we plan to provide material/guidance for a competent Rust programmer to up-skill so that they can teach this course?

I have not considered this, and my inclination is no.



> to be able to apply a position at PingCap(or similar type of positions at other companies) afterwards, or to have a good feeling for better understanding distributed systems (kv stores) and feel much more experienced with Rust afterwards.

@liufuyang yes! the being able to apply for a position at PingCAP is probably PingCAP's primary goal here, though in the README it's understated for a general audience.

Note though that _this_ course doesn't cover distributed systems - it just leads into another course that does.

> About the content/format of the course, some people might like reading docs to learn, but I do like to learn via short videos. I don't know, but perhaps we can even later make this course on platforms such as coursera/edx/udemy?

Yeah, videos are definitely on the roadmap. And it is certainly a challenging task to do them well. Who knows if we'll ever get that far.

Thank you very much @liufuyang for the extensive feedback.




> I think now we have a big ambition now. :-) This course may be hard for students to finish it in a short time. Maybe we need to separate it into different levels like basic and advanced levels. And at first, I think we can only provide a basic level to let students finish all in less than one month.
> 
> Another thing I worry about is that this course may be also difficult for us to make the materials, especially for giving projects, I think we need the communities to help us :-)

It is definitely going to be hard to finish in a short time. I don't expect that it can be finished in a short time as outlined here, which is one of the reasons I mentioned above that I'm worried it _only_ works as an online course, and not offline. But it's structured to be like the MIT distributed systems course, which _definitely_ can't be completed in a short time. I'm unclear on the requirements here.

We should discuss what we can cut out of it. I think the final project ends up about where we discussed it should. What do you think? Most of the sections (except for parallelism) are built around the subjects we discussed too (at least file i/o, networking and async, I think...)

We should talk about what can be cut from this - as we implement it and find that it's too ambitious we'll definitely start wanting to cut things. The entire parallelism section could be cut, since a lot of it is a detour that doesn't matter _too_ much for the final project - we could even make the final project single-threaded. I do think a potential pingcap candidate should better understand parallelism of course.

Getting help from the Rust community would definitely help (though it would significantly reduce the time _I_ could personally spend hacking). Once the structure of the course is firmly established and can be broken down into discrete tasks, and the project starts to look serious, I intend to announce it on users.rust-lang.org with a call for participation, elevate awareness of the project, post regular contributor issues to this week in rust.



> You worried that the course can't be given in Q1, don't worry, I think we can make the materials at first.

Well, I'm worried that the material outlined here can't be completed in Q1, but we'll focus on the most important parts, and have _something_ that looks passable in Q1.



>     * What is the exact audience this course caters at? How much previous programming knowledge should people have?

@skade I tried to cover this in the readme, but maybe wasn't specific enough. I'm thinking it can be taken by a novice programmer, perhaps to level up to an intermediate programmer. So "some" programming experience. It will be challenging for a novice, and they would have to follow the readings and put real work in on their own.



>     * Do you think there's a chance to produce an example program that can be used for teaching concurrency styles that can be shared? I've had great success with sketch programs of certain patterns, but haven't quite found a handle of turning this into a shareable artefact.

I do intend for all projects to have example solutions for any questions they pose. If I understand you right though that's not quite what you are thinking, but rather a practical working program that employs multiple viable concurrency (and parallel presumably) programming patterns.

That sounds like a very cool idea. With the outline here, I would expect to progress _through_  multiple programming styles, but not end up with a program that contains them all. e.g. there's a section that writes the project with synchronous networking, then a section that rewrites it with asynchronous.

So offhand I'd say that, if that was your meaning, such an example is out of scope for this course.

Thanks for the feedback @skade. I'm honored.

Ok, I think I've responded to all the feedback so far. I'll think about it and come back tomorrow with a followup patch.

skade commented:
@brson This looks great :). I have multiple pieces of feedback there.

First, joining other discussions: You already said you are inclined not to provide material that allows others to learn teaching this course. I recommend to say _no_ directly. Teacher teaching is an incredibly hard subject and unless you want to tackle that, don't do it. At most, give people the option to contact you and maybe get info out of you in a 1:1 call. This scales much better on small scale.

Considering people: teaching Rust to beginners is possible, but still rough waters. I'd recommend there to expect people to know a little about programming, or provide a reading list of things you expect them to learn before. I know that there's actually an influx of people wanting to learn programming through Rust _because_ they are interested in distribution, so this is an interesting subject to tackle. I definitely recommend being in touch with other Rust teachers on the subject of beginners.

Considering the artifact: I have an example with a simple server with I implement multiple times in different concurrency fashions (single-threaded, threaded, event-based, event-based i/o+per-request threading). Depending on how much you want to lean on the _distributed_ and _networked_ side of programming though, I'm not sure if this even matters for what you plan on doing. My example is open-source and freely available, but I must admit that I have still bailed on documenting how to use it as a teaching object. See the first paragraph :).

overvenus commented:
FYI, the 6.824 raft labs are hosted in [labs6.824] temporarily. It's not ready yet, missing the raft framework (working in progress) and tests.

lab2 raft and lab3 kvraft should be available by the end of this month.

[labs6.824]: https://github.com/overvenus/labs6824

Hoverbear commented:
@jonhoo That's so cool. That would be really appreciated. :)

Hoverbear commented on rust/README.md:
```diff
@@ -0,0 +1,135 @@
+- Build scripts and the interaction between build-time and run-time
+- Insights into inner workings of the language and libraries
+- How to learn what you don't know about Rust and find what you need to find
+- ... and etc.
```
This line is extraneous, you already say "include" at the start.

brson replied:
I'll remove it.

Hoverbear commented on rust/README.md:
```diff
@@ -0,0 +1,135 @@
+
+Non-goals of this course include teaching installation, syntax and other Rust
+basics; teaching basic data structures and algorithms; being a comprehensive
+resource on the Rust language. That information is easily found [elsewhere].
```
Maybe point directly to the book?

brson replied:
Yes, I'll point to the book, and say so in the text.

Hoverbear commented on rust/README.md:
```diff
@@ -0,0 +1,135 @@
+The course can be followed and completed online or delivered offline as a series
+of training sessions, and consists of
+
+- _readings_ - background material that is expected to be read before each
```
I like this, encouraging people to dig into topics as they learn them is teaching good habits.

Hoverbear commented on rust/README.md:
```diff
@@ -0,0 +1,135 @@
+algorithms and I/O.
+
+This course is for programmers who have begun learning Rust, but still need
+guidance in real Rust programming and best practices.
```
```suggestion
guidance in production Rust programming and best practices.
```

Hoverbear commented on rust/notes.md:
```diff
@@ -0,0 +1,54 @@
+- impl trait, the Into<Option<_>> trick
+- Rust history, culture, design principles
+- rustfmt, clippy, configuring both
+- rust 2018?
```
Given that this is the current and stable edition, yes. I think covering *2015* is optional beyond a small mention.

Hoverbear commented on rust/notes.md:
```diff
@@ -0,0 +1,54 @@
+- using RUSTFLAGS
+- debugging
+- profiling
+- how not to write Rust (bad practices from other languages)
```
Important topic!

Hoverbear commented on rust/notes.md:
```diff
@@ -0,0 +1,54 @@
+- https://pdos.csail.mit.edu/6.824/schedule.html
+  - The MIT distributed systems course this course
+    is inspired by and intended to precede
+- https://github.com/brson/rust-three-days-course
```
Maybe add @skade 's original course link?

skade replied:
It's currently under https://github.com/ferrous-sytems/rust-three-days-course.

brson replied:
Oops. I'll change this.

jonathan-zageno commented:
Looking forward to this!

jonathan-s commented:
Looking forward to this!

jonhoo commented:
I think this looks pretty good! Some notes:

 - I found the presence of "Whirlwind Rust" somewhat strange, as it seems like something that is already assumed background knowledge? I guess this doubles as an "intro to the course", so any content will only be covered relatively briefly.
 - I think it's probably not worth covering padding, alignment, and packing in the data structures lecture. That doesn't become important knowledge until much much later. Enum implementation and optimizations is just on the border-line, but it still seems a little out-of-order given how basic the surrounding lectures are.
 - Rust tooling re-covers clippy and rustfmt?
 - File I/O doesn't seem to really cover anything about File I/O? Collections and iterators also seems like it belongs in the previous section?
 - What is the rationale for covering "build-time Rust" in the networking section? Does GRPC/prost use it in a way that the user has to interact with?
 - In the profiling lesson, `perf` should probably be covered. Maybe also [`bpftrace`](https://www.joyfulbikeshedding.com/blog/2019-01-31-full-system-dynamic-tracing-on-linux-using-ebpf-and-bpftrace.html)? As for blocking profiles, you can probably start [here](http://www.brendangregg.com/offcpuanalysis.html).
 - "The big problem" should probably cover `Send` and `Sync`. Another good resource link might be [this one](https://cglab.ca/~abeinges/blah/too-many-lists/book/). I also found it a little odd that this lesson comes second, not first?
 - "Ownership and borrowing in practice" seems like it should come much sooner?
 - I think there's *a lot* to cover in "Parallel Rust" (things like `thread::spawn`, `rayon`, etc.). It may be tricky to fit that into a single lecture.
 - For "basic futures", my [video lecture](https://www.youtube.com/watch?v=9_3krAQtD2k) might be a good resource, but I'll leave that up to you. It also goes into how that all interacts with Tokio. Maybe link to particular timestamps for the three lectures rather than ask them to watch it all in one go?
 - For "Async I/O", there's now a lot of good documentation in the [Tokio docs](https://tokio.rs/docs/getting-started/hello-world/) that goes quite deep into how things work. I'd probably start with [`mio`](https://github.com/carllerche/mio/) though before moving on to Tokio (assuming the students want to understand how this stuff works too).

brson commented:
OK, after some delay, again thank all for the feedback.

I pushed a commit that addresses as much of the feedback I feel comfortable dealing with in this PR. In particular it

- makes the language more plain in the places requested,
- raises the pre-requisites: "This book is not for novice programmers, but it is for novice Rust programmers." Also, requesting the entire Rust book be read ahead of time.
- added a readme section making clear that instructors will get no guidance within, and that at present they should not attempt to teach this material
- adds various rough info to notes.md
  - fixes the rust-three-days link
  - sections that can be easily cut
- adds various ideas and readings to the plan outline


> First, joining other discussions: You already said you are inclined not to provide material that allows others to learn teaching this course. I recommend to say _no_ directly.

@skade added to the README.


> I'd recommend there to expect people to know a little about programming, or provide a reading list of things you expect them to learn before.

@skade I've strengthened the prerequisites: "This book is not for novice programmers, but it is for novice Rust programmers." It also makes the entire Rust book a pre-requisite.

> Considering the artifact: I have an example with a simple server with I implement multiple times in different concurrency fashions (single-threaded, threaded, event-based, event-based i/o+per-request threading). Depending on how much you want to lean on the _distributed_ and _networked_ side of programming though, I'm not sure if this even matters for what you plan on doing. My example is open-source and freely available, but I must admit that I have still bailed on documenting how to use it as a teaching object. See the first paragraph :).

@skade This sounds awesome, and very much related to what we're trying to do here. I searched for it but didn't find it. Can you provide a link? There's probably a lot to be learned from it, as with rust-three-days. At the least we can link to it at some point.



> FYI, the 6.824 raft labs are hosted in [labs6.824](https://github.com/overvenus/labs6824) temporarily. It's not ready yet, missing the raft framework and tests, done.

@overvenus thank you. I've taken a todo to file an issue to compare your work with mine, make a plan to converge on a common format.


>     * I found the presence of "Whirlwind Rust" somewhat strange, as it seems like something that is already assumed background knowledge? I guess this doubles as an "intro to the course", so any content will only be covered relatively briefly.

@jonhoo thank you very much for taking a close read and offering specific feedback. It's greatly appreciated.

I agree that the subject matter is incongruously basic, particularly now that I've pushed changes to ask students to read the entire Rust Book ahead of time. I was seeing this as an opportunity to do things like: explain the course, the pre-requisites, provided url-shortened-links to various resources, establish rapport. These are admittedly relevant to live instruction more than online.

As some of the most common feedback is about the large scope, I'm looking for things to cut. And while I haven't cut it from this PR, I've added it to a list in notes.md. I'll file a scope-cutting issue after this PR.

>     * I think it's probably not worth covering padding, alignment, and packing in the data structures lecture. That doesn't become important knowledge until much much later. Enum implementation and optimizations is just on the border-line, but it still seems a little out-of-order given how basic the surrounding lectures are.

All of these things are simply interesting, potentially deep, topics related to the subjects of data structures. They don't contribute to the project at hand. They are examples of ideas I came up with for adding value to the lectures beyond the readings.

Furthermore, the entire "data structures in Rust" lesson is motivated because the section 1 project calls for picking a data structure, which in this project should just be something like `struct MyKVStore(HashMap)`. Of course, as the projects increase in scope, more about data structures will need to be known, but still probably nothing beyond what's in the book.

This is a good example of the tension here about providing value beyond the background reading. The entire lesson can be cut, if we're just aiming to solve the project at hand. If we are instead trying to provide expert-level knowledge while also pursuing a goal, then I think there is plenty to say about Rust data structures.

I'll think about this further in a follow-up issue.

> Rust tooling re-covers clippy and rustfmt?

I don't see them covered previously in the lesson plan. Though there may be confusion here because the project is listed before the lessons, and mentions both as part of the topics of the project. The lesson teaches clippy and rustfmt to the extent necessary for the project, and again for the sake of adding value, probably lots more about practical clippy and rustfmt usage.



>     * File I/O doesn't seem to really cover anything about File I/O? Collections and iterators also seems like it belongs in the previous section?

Yeah, it reads pretty weird right now, particularly without some obvious lessons filled in. The sections are all named after the major subject I've previously agreed the course project would be structured around. So here the section and project are centered around file I/O.

The lessons _should_ cover topics necessary to complete the project. That's why the error handling lesson is here - this project will require code that needs error handling (though also section 1 was getting big).

With improved clarity about the scope and audience of this course, there may not be any place for 'collections and iterators'. I was looking for a place to fit it in because a) this course does use both, b) it's a typical "intro to rust" subject.

I didn't have any particular ideas about extra insight into them. I've put it on the list of lessons to cut.

This section should have lessons related to std file I/O and serialization, probably something on write-ahead-logs, which are used in this project.

Needs work.



>     * What is the rationale for covering "build-time Rust" in the networking section? Does GRPC/prost use it in a way that the user has to interact with?

The entire reason for doing it in this section is that using GRPC is going to involve using a build-time protocol compiler. Build-time though is a subject I think that holds a lot of mystery for Rust programmers, and obvious place to provide valuable knowledge that, while not relevant to finishing the project, I expect would hold a lot of interest for budding Rust programmers.

The rationale is related to questions of scope similarly to mention above.


> * In the profiling lesson, `perf` should probably be covered. Maybe also [`bpftrace`](https://www.joyfulbikeshedding.com/blog/2019-01-31-full-system-dynamic-tracing-on-linux-using-ebpf-and-bpftrace.html)? As for blocking profiles, you can probably start [here](http://www.brendangregg.com/offcpuanalysis.html).
> * "The big problem" should probably cover `Send` and `Sync`. Another good resource link might be [this one](https://cglab.ca/~abeinges/blah/too-many-lists/book/). I also found it a little odd that this lesson comes second, not first?

Thanks for the suggestions. I've added them.

As for the lesson ordering, the debugging and profiling subject is high on the list of subjects that could use deeper coverage in the community, that are very useful for real-world programming. I put them in the parallelism section because introducing parallelism is a good opportunity to do some benchmarking exercises.

The choice to put debugging first in the section is pretty arbitrary. I went ahead and put it at the end of the section.

> * "Ownership and borrowing in practice" seems like it should come much sooner?

This is definitely a curious place to put the lesson. Again this is something of a "must have" subject, and I chose to put it here based on the parallelism angle - that unique types is _the_ thing that makes parallelism safe without a GC in Rust; and under the assumption that they've read enough background material to navigate the basic ownership-related tasks required of the projects up to this point. Similarly, parallelism is a great context to introduce the notion of mutable aliasing, the central problem Rust solves.

Unfortunately the entire parallelism section is on the list of things to cut for an mvp, since it doesn't lead directly to the async project. Some of these topics though seem like "must haves" and might need to be wedged into other sections.

>     * I think there's _a lot_ to cover in "Parallel Rust" (things like `thread::spawn`, `rayon`, etc.). It may be tricky to fit that into a single lecture.

You are right about that.

For now I've left the plan alone w/r/t the above feedback, but the problems are noted. We'll see how things shake out later.




> * 
> 
> * For "basic futures", my [video lecture](https://www.youtube.com/watch?v=9_3krAQtD2k) might be a good resource, but I'll leave that up to you. It also goes into how that all interacts with Tokio. Maybe link to particular timestamps for the three lectures rather than ask them to watch it all in one go?
> 
> * For "Async I/O", there's now a lot of good documentation in the [Tokio docs](https://tokio.rs/docs/getting-started/hello-world/) that goes quite deep into how things work. I'd probably start with [`mio`](https://github.com/carllerche/mio/) though before moving on to Tokio (assuming the students want to understand how this stuff works too).

I've added the links as background reading.

I've also indicated in the plan the mio should probably be covered. I do find that I am often in disagreement with the pattern of explaining layered architectures from the inside out. e.g. for practical purposes I don't think someone will need to know about mio to complete this course. That's not to say that I don't think mio should be covered - it's a well of interesting subject matter. But I probably wouldn't prioritize it as high as you, and I might circle back to it after covering basic tokio. We'll see how it plays out!

----

Ok, this PR has sat long enough. I am truly grateful for the feedback from everyone. I'm going to merge it as it stands now with the understanding that this is all subject to change (it's the first meaningful commit to the project). My next steps are: post various follow-up issues; read overvenus's work on the distributed systems examples; post a PR proposing the structure of lessons/projects/examples/tests.

After there's sufficient structure in-repo for others to crib off of, then we'll file issues that others can help with.

I'm always looking for ideas for structuring projects for contributors, and happy to hear any ideas. Maybe not on this PR though since it's abut to be closed.

💖
